### ISSUE TITLE
Crashing against current Master - Broken Pipe

### ISSUE THREAD

**Issue by nritholtz** [thoughtbot/capybara-webkit#777]
[Webkit Debug output](https://gist.github.com/nritholtz/29fefeafc25850ce1ca1) of crashing tests when ran against lastest commit, 7bedbc050d, in master. As you can tell, the tests are running against a remote server.

All system information is included in the Webkit Debug output.

**Comment by jferris** [comment]
It seems likely that we're trying to free an already-deleted reference.

@nritholtz does this reliably crash for you on master but reliably work on a previous commit? If so, is there any way you could pinpoint the commit? The commit you linked to simply updated a NEWS file, and it seems unlikely for that to be the culprit.

Narrowing it down to a more specific test case would also be very valuable.

**Comment by nritholtz** [comment]
I'll try to run it against previous versions, since it is failing consistently on a specific test.

Unfortunately, it will take a little while since this is a bump upgrade from 1.3.0.

I will let you know.

**Comment by nritholtz** [comment]
@jferris f8f2048 is the offending commit
Works fine against previous commit, e49176b

**Comment by jferris** [comment]
Thanks. The commit that's causing issues is fortunately pretty small, so we may be able to debug with a few rounds.

I remember an earlier issue where the logger was causing issues because the URL string was somehow deleted. Based on that, I tried disabling just the logger change from that commit and pushed to a branch called `jf-reset-crash`.

Could you give that branch a shot?

**Comment by nritholtz** [comment]
@jferris Just tried it against your new commit but unfortunately [it crashed again](https://gist.github.com/nritholtz/a08645ef34ade7f11f2d).

**Comment by jferris** [comment]
@nritholtz hmm, okay. I pushed another commit with some debugging statements in the `reset` statement. Can you give that a shot? That will hopefully tell us where the crash is occurring.

**Comment by nritholtz** [comment]
@jferris Tell me if you want the entire debug log, but here is what I think you want as the relevant log:
```
Received "Reset()"
Started "Reset()"
Beginning reset
Pending replies: 1
Aborting pending replies
  	Portal User: auto1431657538
      end of file reached (EOFError)
      /home/qa/.rvm/gems/ruby-2.1.0@automated_testing/bundler/gems/capybara-webkit-9240ce857f8c/lib/capybara/webkit/connection.rb:55:in `read_nonblock'
      /home/qa/.rvm/gems/ruby-2.1.0@automated_testing/bundler/gems/capybara-webkit-9240ce857f8c/lib/capybara/webkit/connection.rb:55:in `read'
      /home/qa/.rvm/gems/ruby-2.1.0@automated_testing/bundler/gems/capybara-webkit-9240ce857f8c/lib/capybara/webkit/connection.rb:46:in `gets'
      /home/qa/.rvm/gems/ruby-2.1.0@automated_testing/bundler/gems/capybara-webkit-9240ce857f8c/lib/capybara/webkit/browser.rb:287:in `check'
      /home/qa/.rvm/gems/ruby-2.1.0@automated_testing/bundler/gems/capybara-webkit-9240ce857f8c/lib/capybara/webkit/browser.rb:210:in `command'
      /home/qa/.rvm/gems/ruby-2.1.0@automated_testing/bundler/gems/capybara-webkit-9240ce857f8c/lib/capybara/webkit/browser.rb:38:in `reset!'
      /home/qa/.rvm/gems/ruby-2.1.0@automated_testing/bundler/gems/capybara-webkit-9240ce857f8c/lib/capybara/webkit/driver.rb:254:in `reset!'
      /home/qa/.rvm/gems/ruby-2.1.0@automated_testing/gems/capybara-2.4.1/lib/capybara/session.rb:103:in `reset!'
      /home/qa/.rvm/gems/ruby-2.1.0@automated_testing/gems/capybara-2.4.1/lib/capybara.rb:257:in `block in reset_sessions!'
      /home/qa/.rvm/gems/ruby-2.1.0@automated_testing/gems/capybara-2.4.1/lib/capybara.rb:257:in `each'
      /home/qa/.rvm/gems/ruby-2.1.0@automated_testing/gems/capybara-2.4.1/lib/capybara.rb:257:in `reset_sessions!'
      /home/qa/.rvm/gems/ruby-2.1.0@automated_testing/gems/capybara-2.4.1/lib/capybara/cucumber.rb:8:in `After'
```

**Comment by jferris** [comment]
@nritholtz thanks. That confirms my suspicion. I'll see if I can replicate this locally and push up a fix later this week.

**Comment by mhoran** [comment]
@jferris, perhaps it would make sense to listen to the `destroyed()` signal from the reply? Maybe some of these replies are getting aborted elsewhere, or not finishing for some reason -- and we have a stale reference?

**Comment by jferris** [comment]
> Perhaps it would make sense to listen to the `destroyed()` signal from the reply?

That sounds like a good idea. It's definitely the `abort` call that's crashing it, so it must be that the replies are getting deleted without emitting `finished`.

**Comment by jwg2s** [comment]
Is the recommended short-term solution here to downgrade to 1.5.0?

**Comment by jferris** [comment]
@jwg2s that will probably work for you. The only change in 1.5.1 was a fix for a crash in 1.5.0, but it unfortunately introduced a new crash. Hopefully you won't hit the other crash on 1.5.0. I'll try to get a fix for this out sometime this week.

**Comment by jwg2s** [comment]
Downgrading worked for us, will look for the next update and then upgrade again.  Thanks for the help!

**Comment by jferris** [comment]
@nritholtz could you try the latest commit on that branch? I think I've managed to reproduce the crash.

**Comment by vkatz-ninthdecimal** [comment]
@jferris, we have had sporadic broken pipe issues in our tests for months now.  i just tried your recent commit, and it looks like it may have fixed them!

**Comment by nritholtz** [comment]
Tried it a few times and it looks good.

**Comment by vkatz-ninthdecimal** [comment]
Hrm, actually i was wrong.  we are still getting broken pipe issues

**Comment by vkatz-ninthdecimal** [comment]
@jferris, should i open a new issue or add stuff here?

**Comment by jferris** [comment]
@vkatz-ninthdecimal if you could open a new issue, that would be great.

**Comment by nritholtz** [comment]
Hey @jferris,

Any update on getting this into a release?

**Comment by jferris** [comment]
I was running into some issues on CI, and I haven't had a chance to address it. My next likely opportunity to work on this is Friday.

**Comment by jferris** [comment]
Fixed by #781. Will be released in 1.5.2.
